internal/dataplane/parser/translators/httproute_atc.go

```diff
@@ -227,3 +229,370 @@ func matchersFromParentHTTPRoute(hostnames []string, metaAnnotations map[string]
+	// split HTTPRoutes by hostname.
+	hostnamedRoutes := []*gatewayv1beta1.HTTPRoute{}
+	if len(httproute.Spec.Hostnames) == 0 {
+		hostnamedRoutes = append(hostnamedRoutes, httproute.DeepCopy())
+	} else {
+		for _, hostname := range httproute.Spec.Hostnames {
+			hostNamedRoute := httproute.DeepCopy()
+			hostNamedRoute.Spec.Hostnames = []gatewayv1beta1.Hostname{hostname}
+			hostnamedRoutes = append(hostnamedRoutes, hostNamedRoute)
+		}
+	}
+	// split HTTPRoutes (already split once by hostname) by match.
+	newHTTPRoutes := []*gatewayv1beta1.HTTPRoute{}
+	for _, route := range hostnamedRoutes {
+		for i, rule := range route.Spec.Rules {
+			for j, match := range rule.Matches {
+				splitRoute := route.DeepCopy()
+				splitRoute.Spec.Rules = []gatewayv1beta1.HTTPRouteRule{
+					{
+						Matches:     []gatewayv1beta1.HTTPRouteMatch{match},
+						Filters:     rule.Filters,
+						BackendRefs: rule.BackendRefs,
+					},
+				}
+				if splitRoute.Annotations == nil {
+					splitRoute.Annotations = map[string]string{}
+				}
+				splitRoute.Annotations[InternalRuleIndexAnnotationKey] = strconv.Itoa(i)
+				splitRoute.Annotations[InternalMatchIndexAnnotationKey] = strconv.Itoa(j)
+				newHTTPRoutes = append(newHTTPRoutes, splitRoute)
+			}
+		}
+	}
+
+	return newHTTPRoutes
+}
+
+type SplitHTTPRouteToKongRoutePriority struct {
+	HTTPRoute *gatewayv1beta1.HTTPRoute
+	Priority  int
+}
+
+type HTTPRoutePriorityTraits struct {
+	PreciseHostname bool
+	HostnameLength  int
+	PathType        gatewayv1beta1.PathMatchType
+	PathLength      int
+	HeaderCount     int
+	HasMethodMatch  bool
+	QueryParamCount int
+}
+
+// CalculateHTTPRoutePriorityTraits calculates the parts of priority that can be decided by the
+// fields in spec of the HTTPRoute. Specification of priority goes as follow:
+// (The following comments are extracted from gateway API specification about HTTPRoute)
+//
+// In the event that multiple HTTPRoutes specify intersecting hostnames,
+// precedence must be given to rules from the HTTPRoute with the largest number of:
+//
+//   - Characters in a matching non-wildcard hostname.
+//   - Characters in a matching hostname.
+//
+// If ties exist across multiple Routes, the matching precedence rules for HTTPRouteMatches takes over.
+//
+// Proxy or Load Balancer routing configuration generated from HTTPRoutes MUST prioritize matches based on the following criteria, continuing on ties.
+// Across all rules specified on applicable Routes, precedence must be given to the match having:
+//
+//   - "Exact” path match.
+//   - "Prefix" path match with largest number of characters.
+//   - Method match.
+//   - Largest number of header matches.
+//   - Largest number of query param matches.
+func CalculateHTTPRoutePriorityTraits(httpRoute *gatewayv1beta1.HTTPRoute) HTTPRoutePriorityTraits {
+	traits := HTTPRoutePriorityTraits{}
+	if len(httpRoute.Spec.Hostnames) != 0 {
+		hostname := httpRoute.Spec.Hostnames[0]
```

Comment:
It would be good to add a comment here about having only one hostname at most, as the `HTTPRoute` has been split. I think it would help future readers. WDYT?

Reply:
added comments.